src/xb-database-manager.c

```diff
@@ -43,157 +42,115 @@
-    gobject_class->finalize = xb_database_manager_finalize;
+  g_slice_free (DatabasePayload, payload);
 }
 
-static void
-xb_database_manager_init (XbDatabaseManager *self)
+static DatabasePayload *
+database_payload_new (XapianDatabase *db,
+                      XapianQueryParser *qp,
+                      GFileMonitor *monitor,
+                      const gchar *lang)
 {
-  XbDatabaseManagerPrivate *priv = xb_database_manager_get_instance_private (self);
+  DatabasePayload *payload;
 
-  priv->stemmers = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_object_unref);
-  g_hash_table_insert (priv->stemmers, g_strdup ("none"), xapian_stem_new ());
+  payload = g_slice_new0 (DatabasePayload);
+  payload->db = g_object_ref (db);
+  payload->qp = g_object_ref (qp);
+  payload->monitor = g_object_ref (monitor);
+  payload->lang = g_strdup (lang);
 
-  priv->database_langs = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
-  priv->databases = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_object_unref);
-  priv->database_qps = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_object_unref);
+  return payload;
+}
 
-  /* Setup the meta database, which will let us query all databases via
-   * the query_all() method
-   */
-  priv->meta_db = xb_database_manager_new_meta_db (self, NULL);
+typedef struct {
+  /* string path => struct DatabasePayload */
+  GHashTable *databases;
+  /* string lang_name => object XapianStem */
+  GHashTable *stemmers;
+} XbDatabaseManagerPrivate;
 
-  /* The prefix store manages stored field -> prefix associations, and
-   * returns unions of unique associations for use in meta databases
-   */
-  priv->prefix_store = xb_prefix_store_new ();
-}
+G_DEFINE_TYPE_WITH_PRIVATE (XbDatabaseManager, xb_database_manager, G_TYPE_OBJECT)
 
-/* Returns whether the manager has a database at index_name */
-gboolean
-xb_database_manager_has_db (XbDatabaseManager *self,
-                            const gchar *index_name)
+static void
+xb_database_manager_invalidate_db (XbDatabaseManager *self,
+                                   const gchar *path)
 {
   XbDatabaseManagerPrivate *priv = xb_database_manager_get_instance_private (self);
-  return (g_hash_table_lookup (priv->databases, index_name) != NULL);
+  g_hash_table_remove (priv->databases, path);
 }
 
-static gboolean
-stem_supports_language (const gchar *lang)
+static void
+database_monitor_changed (GFileMonitor *monitor,
+                          GFile *file,
+                          GFile *other_file,
+                          GFileMonitorEvent event_type,
+                          gpointer user_data)
 {
-  gchar **available_langs;
-  gint idx;
-  gboolean retval = FALSE;
-  const gchar *xapian_lang = NULL;
+  XbDatabaseManager *self = user_data;
+  gchar *path;
 
-  /* Mapping from ISO 639 lang codes to
-   * Xapian's internal language strings.
+  /* GFileMonitor reports "changed" events on a directory's children, but not
```

Comment:
Yep, struct it is. Another, even worse, way is to use `g_object_set_data(database_manager, "payload", payload)` and pass only the database manager as user data.